modules/programs/joplin-desktop.nix

```diff
@@ -0,0 +1,129 @@
+{ config, lib, pkgs, ... }:
+let
+  cfg = config.programs.joplin-desktop;
+  # config path is the same for linux and mac
+  configPath = (config.home.homeDirectory + "/.config/joplin-desktop");
+in {
+  options.programs.joplin-desktop = {
+    enable = lib.mkEnableOption "joplin-desktop";
+    package = lib.mkOption {
+      type = lib.types.package;
+      default = pkgs.joplin-desktop;
+      defaultText = lib.literalExpression "pkgs.joplin-desktop";
+      description = "Package containing the joplin-desktop program.";
+    };
+    extraConfig = lib.mkOption {
+      type = lib.types.attrs;
+      default = { };
+      description = ''
+        Use this to add options to the Joplin config file. Settings are written in JSON, so "sync.interval":600 is written as "sync.interval" = 600'';
+      example = {
+        "newNoteFocus" = "title";
+        "markdown.plugin.mark" = true;
+      };
+    };
+    ### General
+    general = {
+      editor = lib.mkOption {
+        type = lib.types.nullOr lib.types.str;
+        default = null;
+        example = "kate";
+        description =
+          "The editor command (may include arguments) that will be used to open a note. If none is provided Joplin will try to auto-detect the default editor.";
+      };
+    };
+    ### Sync
+    sync = {
+      target = lib.mkOption {
+        type = lib.types.enum [
+          null
+          "none"
+          "file-system"
+          "onedrive"
+          "nextcloud"
+          "webdav"
+          "dropbox"
+          "s3"
+          "joplin-server"
+          "joplin-cloud"
+        ];
+        default = null;
+        description = "What is the type of sync target.";
+        example = "dropbox";
+      };
+      interval = lib.mkOption {
+        type = lib.types.nullOr lib.types.int;
+        default = null;
+        description = ''
+          Sync Interval in seconds. Only values shown as options in the settings are valid.
+          Disabled: 0
+          5 Min: 300
+          10 Min: 600
+          30 Min: 1800
+          1 hour: 3600
+          12 Hours: 43200
+          24 Hours: 86400
+        '';
+        example = 600;
+      };
+    };
+  };
+
+  config = lib.mkIf cfg.enable {
+    home.packages = [ cfg.package ];
+
+    # write new config to temporary read-only file
+    home.file."home-manager-settings.json" = {
+      target = (configPath + "/home-manager-settings.json");
+      text = builtins.toJSON ((
```

Comment:
I think this is best done using existing Home Manager functionality:

``` nix
xdg.configFile."joplin-desktop/settings.json".source = jsonFormat.generate "joplin-settings.json" ((
```

where

``` nix
jsonFormat = pkgs.formats.json { };
```

is defined at the top of the file, after `cfg = config.programs.joplin-desktop;`.

If you really do need to merge the HM generated configuration into the one generated by Joplin Desktop, then you don't need to put anything in the user's home directory. You will instead need use `jq` in the activation block and reference the HM generated file directly in the Nix store.

Reply:
Yes it has to be merged because Joplin writes some other things in that file. I'll rewrite it to use `jq` and the file from the nix store.

Reply:
I've changed the way the config file is written to use `jq` like you suggested, see latest commit.